appmap/_implementation/recorder.py

```diff
@@ -1,81 +1,184 @@
 import logging
 import threading
 import traceback
+from abc import ABC, abstractmethod
 
-from .event import Event, _EventIds
-from .metadata import Metadata
+from appmap._implementation.utils import appmap_tls
 
 logger = logging.getLogger(__name__)
 
+_default_recorder = None
 
-class Recorder:
 
-    # _recorders is a dict of the in-progress recordings. It's keys are an thread-specific
-    # recorders, and None if a global recorder was created.
-    _recorders = None
-    _lock = threading.Lock()
+class Recorder(ABC):
+    """
+    A base class for Recorders.
+
+    Note that the abstract methods have implementations for use by subclasses.
+    """
+
+    @property
+    @abstractmethod
+    def events(self):
+        return self._events
+
+    @property
+    @abstractmethod
+    def next_event_id(self):
+        self._next_event_id += 1
+        return self._next_event_id
+
+    # It might be nice to put @property on the getters here. The python maintainers have gone back
+    # and forth on whether you should be able to combine @classmethod and @property. In 3.11,
+    # they've decided you can't: https://docs.python.org/3.11/library/functions.html#classmethod.
+    @classmethod
+    def get_current(cls):
```

Comment:
How about making this `__new__` instead? This way you could do `Recorder()` instead of `Recorder.get_current()`. Or would it be too confusing? :)